### PR TITLE
[Merged by Bors] - feat: set compression algorithm behind feature flag

### PR DESCRIPTION
**Pull request by EstebanBorai**

### CODE REVIEW

File: crates/fluvio-compression/Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "fluvio-compression"
-version = "0.3.0"
```
Comment (morenol) [inline crates/fluvio-compression/Cargo.toml:3]: I think that we haven't released 0.3.0, so this bump is not needed

File: crates/fluvio-compression/Cargo.toml [hunk 2]
```diff
@@ -10,13 +10,16 @@ repository = "https://github.com/infinyon/fluvio"
...
 
 [features]
-default = []
+default = ["compress"]
+compress = ["dep:flate2", "dep:lz4_flex", "dep:zstd"]
 
 [dependencies]
 bytes = { workspace = true }
-flate2 = { workspace = true }
 snap = { version = "1" }
```
Comment (morenol) [inline crates/fluvio-compression/Cargo.toml:18]: this also should be optional

Comment (EstebanBorai) [reply crates/fluvio-compression/Cargo.toml:18]: Yeah! I did on: a2390cb

**Comment by EstebanBorai** [comment]
I started to find issues with disengaging compression from records here:

```rust
impl TryFrom<Batch> for Batch<RawRecords> {
    type Error = CompressionError;
    fn try_from(f: Batch) -> Result<Self, Self::Error> {
        let mut buf = Vec::new();
        f.records.encode(&mut buf, 0)?;

        let compression = f.get_compression()?;
        let compressed_records = compression.compress(&buf)?;
        let compressed_records_len = compressed_records.len() as i32;
        let records = RawRecords(compressed_records);

        Ok(Batch {
            base_offset: f.base_offset,
            batch_len: compressed_records_len,
            header: f.header,
            records,
        })
    }
}
```

Perhaps zstd could be an optional feature?

What do you think @galibey, @morenol?

**Comment by morenol** [comment]
is that `try_from` used on SmartEngine?

File: crates/fluvio-compression/Cargo.toml (hunk 2 above)
Comment (morenol) [reply crates/fluvio-compression/Cargo.toml:18]: I mean, the `snap` crate.

File: crates/fluvio-compression/src/error.rs [hunk 3]
```diff
@@ -6,12 +6,13 @@ use snap::write::{IntoInnerError, FrameEncoder};
 pub enum CompressionError {
     #[error(transparent)]
     IoError(#[from] std::io::Error),
+    #[error("Unreachable error")]
+    UnreachableError,
```
Comment (galibey) [inline crates/fluvio-compression/src/error.rs:10]: Why is this error added?

**Comment by galibey** [comment]
> I started to find issues with disengaging compression from records here:
> 
> ```rust
> impl TryFrom<Batch> for Batch<RawRecords> {
>     type Error = CompressionError;
>     fn try_from(f: Batch) -> Result<Self, Self::Error> {
>         let mut buf = Vec::new();
>         f.records.encode(&mut buf, 0)?;
> 
>         let compression = f.get_compression()?;
>         let compressed_records = compression.compress(&buf)?;
>         let compressed_records_len = compressed_records.len() as i32;
>         let records = RawRecords(compressed_records);
> 
>         Ok(Batch {
>             base_offset: f.base_offset,
>             batch_len: compressed_records_len,
>             header: f.header,
>             records,
>         })
>     }
> }
> ```
> 
> Perhaps zstd could be an optional feature?
> 
> What do you think @galibey, @morenol?

Not sure I understood the question, but you can always guard this impl block by feature, right?

**Comment by EstebanBorai** [comment]
> Not sure I understood the question, but you can always guard this impl block by feature, right?

Thats right!

> I think so, that block and some other blocks that are using it can be feature flagged

Im on it!

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -928,7 +928,7 @@ jobs:
     strategy:
  #     fail-fast: false
       matrix:
-        os: [ubuntu-latest]
+        os: [ubuntu-latest, macos-latest]
```
Comment (morenol) [inline .github/workflows/ci.yml:931]: is this needed? We do replicate the issue with the code on `master`? if that is not the case, I think that we should not add this

Comment (EstebanBorai) [reply .github/workflows/ci.yml:931]: I added it to check on compilation issues with macos, by building `smartmodules/examples`

Comment (morenol) [reply .github/workflows/ci.yml:931]: Yes, but did you replicate the issue with master? I see that CI for macos is failing due to changes not related to this PR

Comment (EstebanBorai) [reply .github/workflows/ci.yml]: You are right! Applied [here](https://github.com/infinyon/fluvio/pull/3275/commits/a01add23578c349133adcb2d6557e9484d8a56a7)

**Review by morenol — APPROVED** [review]
LGTM

File: crates/fluvio-compression/Cargo.toml [hunk 5]
```diff
@@ -10,13 +10,16 @@ repository = "https://github.com/infinyon/fluvio"
 description = "Fluvio Compression library"
 
 [features]
-default = []
+default = ["compress"]
+compress = ["flate2", "lz4_flex", "snap", "zstd"]
 
 [dependencies]
 bytes = { workspace = true }
```
[the remark this answers is not in the document]
Comment (EstebanBorai) [reply crates/fluvio-compression/Cargo.toml]: Applied on https://github.com/infinyon/fluvio/pull/3275/commits/86de3059e2f3f13cc35ae652f219bde521b1a38d

**Review by galibey — APPROVED** [review]

**Pull request closed without merging**